cairo-contracts/protobuf/src/types/message.cairo

```diff
@@ -0,0 +1,140 @@
+use protobuf::types::tag::{WireType, ProtobufTag, ProtobufTagImpl};
+use protobuf::primitives::numeric::NumberAsProtoMessage;
+use protobuf::types::wkt::Any;
+
+pub trait ProtoMessage<T> {
+    fn decode_raw(ref value: T, serialized: @ByteArray, ref index: usize, length: usize);
+    fn encode_raw(self: @T, ref output: ByteArray);
+    fn wire_type() -> WireType;
+    fn type_url() -> ByteArray;
+}
+
+pub struct ProtoCodec {}
+
+#[generate_trait]
+pub impl ProtoCodecImpl of ProtoCodecTrait {
+    fn decode<T, +ProtoMessage<T>, +Drop<T>, +Default<T>>(serialized: @ByteArray) -> T {
+        let mut index = 0;
+        let length = if ProtoMessage::<T>::wire_type() == WireType::LengthDelimited {
+            serialized.len()
+        } else {
+            0
+        };
+        let mut value = Default::<T>::default();
+        ProtoMessage::<T>::decode_raw(ref value, serialized, ref index, length);
+        value
+    }
+
+    fn encode<T, +ProtoMessage<T>>(value: @T) -> ByteArray {
+        let mut bytes = "";
+        value.encode_raw(ref bytes);
+        bytes
+    }
+
+    fn from_any<T, +ProtoMessage<T>, +Drop<T>, +Default<T>>(any: @Any) -> T {
+        if any.type_url != @ProtoMessage::<T>::type_url() {
+            panic!("unexpected type URL");
+        }
+        Self::decode::<T>(any.value)
+    }
+
+    fn to_any<T, +ProtoMessage<T>>(value: @T) -> Any {
+        Any { type_url: ProtoMessage::<T>::type_url(), value: Self::encode(value) }
+    }
+
+
+    fn decode_field<T, +ProtoMessage<T>, +Drop<T>>(
+        field_number: u8, ref value: T, serialized: @ByteArray, ref index: usize, bound: usize
+    ) {
+        assert(bound <= serialized.len(), 'invalid bound');
```

Comment:
It would be good to define an `errors.cairo` to handle possible error messages from there.